robot/web/web_c.go

```diff
@@ -84,150 +82,83 @@ func (svc *webService) streamInitialized() bool {
 
 func (svc *webService) addNewStreams(ctx context.Context) error {
 	if !svc.streamInitialized() {
-		return nil
+		return errors.New("stream server not initialized")
 	}
+
+	// Refreshing sources will walk the robot resources for anything implementing the camera and
+	// audioinput APIs and mutate the `svc.videoSources` and `svc.audioSources` maps.
 	svc.refreshVideoSources()
 	svc.refreshAudioSources()
+
 	if svc.opts.streamConfig == nil {
+		// The `streamConfig` dictates the video and audio encoder libraries to use. We can't do
+		// much if none are present.
 		if len(svc.videoSources) != 0 || len(svc.audioSources) != 0 {
-			svc.logger.Debug("not starting streams due to no stream config being set")
+			svc.logger.Warn("not starting streams due to no stream config being set")
 		}
 		return nil
 	}
 
-	newStream := func(name string, isVideo bool) (gostream.Stream, bool, error) {
-		// Configure new stream
+	for name := range svc.videoSources {
+		// We walk the updated set of `videoSources` and ensure all of the sources are "created" and
+		// "started".
 		config := gostream.StreamConfig{
-			Name: name,
-		}
-
-		if isVideo {
-			config.VideoEncoderFactory = svc.opts.streamConfig.VideoEncoderFactory
-		} else {
-			config.AudioEncoderFactory = svc.opts.streamConfig.AudioEncoderFactory
-		}
-		stream, err := svc.streamServer.Server.NewStream(config)
-
-		// Skip if stream is already registered, otherwise raise any other errors
-		var registeredError *webstream.StreamAlreadyRegisteredError
-		if errors.As(err, &registeredError) {
-			return nil, true, nil
-		} else if err != nil {
-			return nil, false, err
+			Name:                name,
+			VideoEncoderFactory: svc.opts.streamConfig.VideoEncoderFactory,
 		}
 
-		if !svc.streamServer.HasStreams {
-			svc.streamServer.HasStreams = true
-		}
-		return stream, false, nil
-	}
-
-	for name, source := range svc.videoSources {
-		const isVideo = true
-		stream, alreadyRegistered, err := newStream(name, isVideo)
+		// Call `createStream`. `createStream` is responsible for first checking if the stream
+		// already exists.
+		//
+		// TODO(RSDK-9079) Add reliable framerate fetcher for stream videosources
+		stream, err := svc.createStream(config, name)
 		if err != nil {
 			return err
-		} else if alreadyRegistered {
-			continue
 		}
 
-		svc.startVideoStream(ctx, source, stream)
+		// If the stream already exists, `createStream` will return a `nil` stream and `nil`
```

Comment:
@seanavery can you double check this for me? I'm pretty confident I described how `createStream` behaves, but I wasn't convinced `startVideoStream` coped with that.

Reply:
Thanks for flagging this comment. `startVideoStream` does not cope with a nil stream. Added back in `alreadyRegistered` check to manually skip -- will update comment here.